docs: add cluster update behavior (#2134)
spectrocloud/librarium: docs/docs-content/clusters/edge/upgrade-behavior.md

Reviewer: Let's keep this page where it is right now but maybe start thinking where other similar topics will reside in the future. I have a feeling we'll have more of these in the future.

Author: Yeah, I was thinking if we have more upgrade related pages, this could become a folder.

---
sidebar_label: "Edge Cluster Upgrade Behavior"
title: "Edge Cluster Upgrade Behavior"
description: "Learn about how Palette Edge responds to upgrades of the cluster profile."
hide_table_of_contents: false
sidebar_position: 80
tags: ["edge", "architecture"]
---

When you update an active Edge cluster's profile, Palette will upgrade the active cluster to the latest version of the
profile. Depending on the nature of the change, implementing an upgrade might involve repaving a cluster, rebooting a
cluster, restarting services, or doing nothing. For more information about cluster repaves, refer to

> [Repave Behavior and Configurations](../cluster-management/node-pool.md#repave-behavior-and-configuration).

Reviewer: @lennessyy can we link to this page in `/clusters/cluster-management/cluster-updates` - I think that would help with cross reference.

## Upgrade Behaviors

> A cluster could respond to an upgrade in several ways.

Reviewer: ```suggestion
A cluster could respond to an upgrade in several ways. The table below lists the potential upgrade behaviors you could encounter.
```
Maybe make it a bit more clear what the table is for? Your call 😄

| Upgrade Behavior     | Description                                                                                                                                             |
| -------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Repave               | Repaves all nodes in the cluster, starting with the control plane.                                                                                      |

> | Reboot               | Reboots all nodes in the cluster, starting with the control plane                                                                                       |

Reviewer 2: ```suggestion
| Reboot               | Reboots all nodes in the cluster, starting with the control plane.                                                                                       |
```

> | Service restart      | Restarts specific services in all nodes in the cluster, starting with the control plane                                                                 |

Reviewer 2: ```suggestion
| Service restart      | Restarts specific services in all nodes in the cluster, starting with the control plane.                                                                |
```

> | Configuration reload | Updates the configurations without needing to restart services or the node. This is only triggered when you update `stage.reconcile.*` in the OS layer. |

Reviewer: ```suggestion
| Configuration reload | Updates the configurations without needing to restart services or the node. This is only triggered when you update `stage.reconcile.*` in the Operating System (OS) layer. |
```
First mention of OS

> | No-op                | Acknowledges the update request but does nothing.                                                                                                       |

Reviewer: @lennessyy the term `no-op` most tech savy people understand it means no operations but maybe we should explain that in description?

> ## Upgrade Behavior By Change Location

Reviewer 2: ```suggestion
## Upgrade Behavior by Change Location
```

Reviewer: Yup, three words is the minimum requirement for title case in titles.

> Updates in the OS layer or the Kubernetes lay can trigger different upgrade behavior depending on the exact values

Reviewer 2: ```suggestion
Updates in the OS or Kubernetes layers can trigger different upgrade behaviors depending on the exact values
```

Reviewer: Lenny, I think it would help if we mention cloud-init as that is where the stages come from. A novice user could struggle tying the concepts together.

changed.

### OS Layer

> | Repave                          | Reboot                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | Service Restart | No-op                                                                                                                                                                                                                                                                                                       |

Reviewer: Same feedback as above about `no-op` 😅

| <li> `options.system.uri` </li> | <li>`stages.rootfs.*`</li> <li>`stages.initramfs.*`</li> <li>`stages.boot.*`</li> <li>`stages.fs.*`</li> <li>`stages.network.*`</li> <li>`stages.after-install.*`</li> <li>`stages.after-install-chroot.*`</li> <li>`stages.after-upgrade.*`</li> <li>`stages.after-upgrade-chroot.*`</li> <li>`stages.after-reset.*`</li> <li>`stages.after-reset-chroot.*`</li> <li>`stages.before-install.*`</li> <li>`stages.before-upgrade.*`</li> <li>`stages.before-reset.*`</li> | None.           | <li>`pack.*`</li> <li>`providerCredentials.*`</li> <li>`options.system.registry`</li> <li>`options.system.repo`</li> <li>`options.system.k8sDistribution`</li> <li>`options.system.osName`</li> <li>`options.system.peVersion`</li> <li>`options.system.customTag`</li> <li>`options.system.osVersion`</li> |

> ### Kubernetes Layer

Reviewer 2: Do we have the same values for the three tabs? Just to confirm :)

Reviewer: I noticed that too.

Author: yep, they are all the same. The content was provided to me in three tabs, and i made them separate so that in the future if they differ, it's easy to manage

</TabItem>

<TabItem label="RKE2" value="rke2">
| Repave | Reboot | Service Restart | No-op |
|--------|--------|-----------------|-------|
| None. | <ul> <li>`stages.rootfs.*`</li> <li>`stages.initramfs.*`</li> <li>`stages.boot.*`</li> <li>`stages.fs.*`</li> <li>`stages.network.*`</li> <li>`stages.after-install.*`</li> <li>`stages.after-install-chroot.*`</li> <li>`stages.after-upgrade.*`</li> <li>`stages.after-upgrade-chroot.*`</li> <li>`stages.after-reset.*`</li> <li>`stages.after-reset-chroot.*`</li> <li>`stages.before-install.*`</li> <li>`stages.before-upgrade.*`</li> <li>`stages.before-reset.*`</li> </ul> | <li> `cluster.*` </li> | <li> `clientConfig.*` </li> <li>`pack.*` </li><li> Any future new keys</li> |

</TabItem>

</Tabs>

## Network Layer

> Changes made to the same Container Network Interface (CNI) pack typically do not result in cluster repave or reboot, and

Reviewer 2: ```suggestion
Changes made to the Container Network Interface (CNI) pack typically do not result in cluster repave or reboot, and
```

can be applied by restarting the related networking services.

:::warning

Do not change to a different CNI pack after provisioning a cluster. You can make changes to the existing CNI pack, but

> if you want to use a different CNI pack altogether, we recommend you re-create another cluster.

Reviewer 2: re-create another cluster or create another cluster? or re-create the existing cluster? :B

:::

## Storage Layer

> Changes made to the storage layer typically do not result in cluster repave or reboot, and can be applied by restarting

Reviewer 2: ```suggestion
Changes made to the storage layer typically do not result in cluster repave or reboot and can be applied by restarting
```
